### ISSUE TITLE
Suggestion: allow to configure LogLevel based on query duration

### ISSUE THREAD

**Issue by holtkamp** [abacaphiliac/doctrine-psr-sql-logger#4]
First of all: thanks for this library, easy to use 👍 

While using it, an idea popped into mind: 

It might be nice to be able to determine the LogLevel dynamically, based on the query duration. Since the duration is only known when the query has finished, this derived LogLevel only applies for the "Query finished" log entry.

For example by handing over a "configuration array" instead of a fixed log-level:

```php 
$logLevelsForQueryDurationsInMilliseconds = [
  0 => LogLevel::INFO, //default value: used for "Query started" and when no specific LogLevel applies, for this example up to 99 milliseconds
  100 => LogLevel::NOTICE, //Used when a query takes more than 100 milliseconds (0.1 second)
  500 => LogLevel::WARNING,  //Used when a query takes more than 500 milliseconds (0.5 second)
  1000 => LogLevel::CRITICAL,  //Used when a query takes more than 1000 milliseconds (1 second)
];
```

What do you think? I can try to come up with a PR...

**Comment by abacaphiliac** [comment]
i really like this idea, and pull requests are very much welcome.

your idea reminds me of a Slow Query log. i think it would be best to create a new class for this, since the configuration is so different from the other implementations. do you agree?

**Comment by holtkamp** [comment]
> i think it would be best to create a new class for this, since the configuration is so different from the other implementations. do you agree?

Well, I suppose it is the challenge to prevent having to change the current signature / interface 🤓 

We could do something like this:

```PHP

/** @var string */
private $defaultLogLevel;

/** @var ?array */
private $logLevelMapping;

public function __construct(LoggerInterface $logger, string $defaultLogLevel = LogLevel::INFO, ?array $logLevelMapping = null )
{
    $this->defaultLogLevel = $defaultLogLevel;
    $this->logLevelMapping = $logLevelMapping;
}

private function getApplicableLogLevel(float $duration): string
{
    return is_array($this->logLevelMapping)
        ? $this->determineApplicableLogLevel($defaultLevel, $duration);
        : $this->defaultLevel;
}

private function determineApplicableLogLevel(float $duration) : string
{
    $durations = array_keys($this->logLevelMapping);
    $durations[] = $duration; //Append the incoming query duration to the array of duration thresholds
    
    asort($durations, SORT_NUMERIC); //Sort the array from low to high: the provided duration will end up somewhere between the thresholds
    $durations = array_values($durations); //A re-index is required after sorting
    
    $key = array_search($duration, $durations, true); //Determine at which position the duration ended up after sorting
    
    $logLevels = array_values($this->logLevelMapping);
    $logLevel = $logLevels[$key - 1]; //Now take the "previous" key
    
    return $logLevel;
}
```

But while I am writing this pseudocode I realize: having to determine the applicable log level for each different $duration, might have some impact on the performance? So maybe some rounding of the $duration and "local" caching should be applied as well... Not sure, this is something to research a bit... 

You can see this initial idea here: https://3v4l.org/OQJD1

**Comment by abacaphiliac** [comment]
i like where you're going with the pseudo and the ev4l sample for the implementation.

regarding performance and caching, i think we can probably keep this reasonably fast by doing some of the level map restructuring in the constructor. of course, this will only benefit shared instances of the query logger, so it will be up to the application owners to sort out which loggers can be shared between connections. that will be worth documenting.

in consideration of worst case scenarios, i think that we can also keep the level-map small by requiring use of the log levels as keys instead of the durations. the runtime sorting will only be over 9 items in the worst case (8 psr levels, plus the runtime duration). i think this feels more natural. we can provide a default set (if we feel that's appropriate), allowing users to override a finite set of known keys. i fear that some need to override the numeric duration keys will lead to over-specification, causing confusion and frustration.

if we can agree that the level map should use levels for keys, we may want to introduce a configuration object so that we can strictly type. alternatively, we can run some assertions in the constructor.

to me, this feels pretty well contained, and after all is said and done i think we can look into some performance testing to make sure we're both comfortable.

i think that instead of adding a compatible secondary configuration mode and increasing the constructor and runtime complexity (albeit small, for now) of the existing class, we should add a new class specifically for slow query logging and completely separate the two modes. in the end, there may be an opportunity to extract common functionality (formatting, ID state, etc.) but we can explore that at a later time.

**Comment by holtkamp** [comment]
> i think that we can also keep the level-map small by requiring use of the log levels as keys instead of the durations

Ah, yeah, that is a better approach indeed!

> we may want to introduce a configuration object so that we can strictly type

indeed

>we should add a new class specifically for slow query logging and completely separate the two modes

Ok, makes sense indeed to keep it separated

**Comment by abacaphiliac** [comment]
hi @holtkamp, just checking in. are you still interested in [contributing] this feature?

**Comment by holtkamp** [comment]
Hi @abacaphiliac, yeah still seems like a nice feature. Unfortunately my time is scarce and this "nice-to-have" is quite low on my priority list... Hopefully I can spend some time on it in August.

**Comment by holtkamp** [comment]
So August 2019 and even 2019 passed. Did not find time to work on this yet...

**Comment by abacaphiliac** [comment]
@holtkamp it's no problem. are you still thinking about putting some time towards it when you can? if you'd rather, i can take a crack at it and submit to you for review.

**Comment by holtkamp** [comment]
@abacaphiliac an initial step is to use the logLevels as keys:

```php
/** @var string */
private $defaultLogLevel;

/** @var ?array */
private $logLevelMapping;

public function __construct(LoggerInterface $logger, string $defaultLogLevel = LogLevel::INFO, ?array $logLevelMapping = null )
{
    $this->defaultLogLevel = $defaultLogLevel;
    $this->logLevelMapping = $logLevelMapping;
}

private function getApplicableLogLevel(float $duration): string
{
    return is_array($this->logLevelMapping)
        ? $this->determineApplicableLogLevel($duration);
        : $this->defaultLogLevel;
}

private function determineApplicableLogLevel(float $duration) : string
{
    $durations = array_values($this->logLevelMapping); //Acquire a common / non-associative array
    $durations[] = $duration; //Append the incoming query duration to the array of duration thresholds
    
    asort($durations, SORT_NUMERIC); //Sort the array from low to high: the provided duration will end up somewhere between the thresholds
    $durations = array_values($durations); //A re-index is required after sorting
    
    $key = array_search($duration, $durations, true); //Determine at which position the duration ended up after sorting
    
    $logLevels = array_keys($this->logLevelMapping);
    $logLevel = $logLevels[$key - 1]; //Now take the "previous" key
    
    return $logLevel;
}
```

And with an example here: https://3v4l.org/QeLAm

Will try to come up with a PR for this initial simple approach using an array.

**Comment by abacaphiliac** [comment]
resolved by #9
